### PR TITLE
Feature api cli tags

### PR DESCRIPTION
**Pull request by FrancoLiberali**
## Description

Possibility of adding tags to the entities.
Add an image to the tags. This image is sent during the creation/edit of a tag in the same json, in data uri schema format. It is created in the database, in the "images" collection and its id is stored in the tag. When the tag is obtained, a relative path is obtained to retrieve that image (api/images + imageID).

See after https://github.com/ditrit/OGrEE-Core/pull/243 and https://github.com/ditrit/OGrEE-Core/pull/238

Fixes https://github.com/ditrit/OGrEE-Core/issues/101

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How Has This Been Tested?

Automatic tests

### CODE REVIEW

File: API/models/model.go [hunk 1]
```diff
@@ -652,133 +614,175 @@ func DeleteEntity(entity string, id string, userRoles map[string]Role) *u.Error
...
-	if ok, err := ValidateEntity(u.EntityStrToInt(ent), t); !ok {
-		return nil, err
-	}
+	return nil
+}
 
-	t["lastUpdated"] = primitive.NewDateTimeFromTime(time.Now())
-	t["createdDate"] = oldObj["createdDate"]
-	delete(t, "parentId")
+func UpdateObject(entityStr string, id string, updateData map[string]interface{}, isPatch bool, userRoles map[string]Role) (map[string]interface{}, *u.Error) {
```
Comment (helderbetiol) [inline API/models/model.go:695]: We can add tags without creating them first, is it intended to be like this?

File: API/models/schemas/tag_schema.json [hunk 2]
```diff
@@ -0,0 +1,52 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "title": "OGrEE Tag Schema",
+  "type": "object",
+  "properties": {
+    "slug": {
+      "type": "string",
+      "$ref": "refs/types.json#/definitions/slug"
+    },
+    "description": {
```
Comment (helderbetiol) [inline API/models/schemas/tag_schema.json:10]: Description for all others is an array of string and I don't see why, we can check with the team to make it always a string

File: API/models/model.go [hunk 3]
```diff
@@ -652,133 +614,175 @@ func DeleteEntity(entity string, id string, userRoles map[string]Role) *u.Error
...
 		if mongoRes.Err() != nil {
 			return nil, mongoRes.Err()
 		}
-		if oldObj["id"] != t["id"] {
+
+		if oldObj["id"] != updateData["id"] {
 			// Changes to id should be propagated to its children
-			if err := PropagateParentIdChange(ctx, oldObj["id"].(string),
-				t["id"].(string), entInt); err != nil {
+			err := PropagateParentIdChange(
```
Comment (helderbetiol) [inline API/models/model.go:765]: A change to the name of the domain should also propagate to all objects that have it or the name should not be modifiable.

File: API/utils/util.go [hunk 4]
```diff
@@ -35,30 +38,40 @@ const (
...
 const NAME_REGEX = "\\w(\\w|\\-)*" // accepted regex for names that compose ids
 const RESET_TAG = "RESET"          // used as email to identify a reset token
+const HIERARCHYOBJS_ENT = "hierarchy_object"
 
 type RequestFilters struct {
-	FieldsToShow []string `schema:"fieldOnly"`
-	StartDate    string   `schema:"startDate"`
-	EndDate      string   `schema:"endDate"`
-	Limit        string   `schema:"limit"`
+	FieldsToShow []string  `schema:"fieldOnly"`
```
Comment (helderbetiol) [inline API/utils/util.go:69]: When use fieldOnly for /all endpoints, tags are also returned.

File: .github/workflows/api-unit-test.yml [hunk 5]
```diff
@@ -22,30 +22,13 @@ jobs:
     - name: Set up Go
       uses: actions/setup-go@v4
       with:
-        go-version: 1.19
-      
-    - name: Pull MongoDB
-      run: docker pull mongo
+        go-version: 1.21
```
Comment (helderbetiol) [inline .github/workflows/api-unit-test.yml:25]: at go.mod is 1.20

Comment (FrancoLiberali) [reply .github/workflows/api-unit-test.yml:25]: in go.mod what is shown is the minimum version of go required to compile, but any version above that can be used, 1.21 in this case

File: API/models/schemas/tag_schema.json (hunk 2 above)
Comment (FrancoLiberali) [reply API/models/schemas/tag_schema.json:10]: I think that in the rest of the cases it is to be able to put several descriptions. I don't see why we need that in a tag (neither in the others, but I guess they have done it that way for a reason...)

File: API/repository/base.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+	}
+
+	return nil
+}
+
+func SetupDB(db *mongo.Database) error {
+	// Indexes creation
+	// Enforce unique children
+	for _, entity := range []int{u.DOMAIN, u.SITE, u.BLDG, u.ROOM, u.RACK, u.DEVICE, u.AC, u.PWRPNL, u.CABINET, u.CORRIDOR, u.GROUP, u.STRAYOBJ} {
+		if err := createUniqueIndex(db, entity, bson.M{"id": 1}); err != nil {
```
[the remark this answers is not in the document]
Comment (FrancoLiberali) [reply API/repository/base.go:53]: Why can it be here? As you say, it will be executed every time the api is launched, but if the index already exists it doesn't generate any problem, it just ignores the creation.
Why should it be here? Because the index creation is not an operational task of the database independent of the api operation (as it is for example to create the first user). If the indexes do not exist the api does not work correctly, because the only verification that the ids are not repeated is thanks to these indexes. Therefore, the api must make sure that the indices exist to make sure it runs on a base that has the right characteristics to work properly.
Why do I need it to be here? As I am adding integration tests, I need to clean the database between runs. To do this, I drop the database and then reconfigure it, which is much faster than deleting all the elements of a database.

File: deploy/db/createdb.js [hunk 7]
```diff
@@ -78,59 +78,9 @@ authDB.auth(ADMIN_USER, ADMIN_PASS);
 
 // Create a new Database
 var db = m.getDB("ogree"+DB_NAME)
-db.createCollection('account');
```
[the remark this answers is not in the document]
Comment (FrancoLiberali) [reply deploy/db/createdb.js:81]: In mongoDB it is not necessary to create the collections beforehand, when the api creates the first object in a collection, the collection is created automatically

File: API/controllers/delete.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package controllers
+
+import (
+	"p3/models"
+	u "p3/utils"
+)
+
+func deleteObject(entityStr string, id string, userRoles map[string]models.Role) *u.Error {
```
[the remark this answers is not in the document]
Comment (FrancoLiberali) [reply API/controllers/delete.go]: Moved to models in b00ec118

File: API/models/model.go (hunk 1 above)
Comment (FrancoLiberali) [reply API/models/model.go:695]: No, this should not be possible. In models/tag.go:26 it checks for the existence of a tag before adding it. Have you found a way to do this?

File: API/models/model.go (hunk 3 above)
Comment (FrancoLiberali) [reply API/models/model.go:765]: To be honest I don't understand this comment, here I only added enters between the parameters because the line was too long, but I didn't change anything in the logic.

File: API/models/model.go (hunk 1 above)
Comment (helderbetiol) [reply API/models/model.go:695]: Using Postman, I sent a PUT and also a POST with tags:[tag that does not exist] and the object was created/modified with it

File: API/repository/base.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+// Initial data creation
+func createInitialData(db *mongo.Database, dbName string) error {
+	// Create a default domain
+	ctx, cancel := u.Connect()
+	defer cancel()
+
+	_, err := CreateObject(ctx, u.EntityToString(u.DOMAIN), map[string]any{
+		"id":       dbName,
+		"name":     dbName,
```
[the remark this answers is not in the document]
Comment (FrancoLiberali) [reply API/repository/base.go]: Fixed in 7709f31d

File: API/utils/util.go (hunk 4 above)
Comment (FrancoLiberali) [reply API/utils/util.go:69]: Fixed in e0add398

File: API/models/model.go (hunk 3 above)
Comment (helderbetiol) [reply API/models/model.go:768]: It was something I noticed testing this branch but it was already there, I put the comment to not forget it, I'll create a separate issue for it

File: API/repository/base.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (helderbetiol) [reply API/repository/base.go:53]: Both approaches seem ok to me, because createdb.js is still mandatory in any case, it should always be run before the API and just once, so the idea of creating collections and indexes there (which should always be done only once) makes sense (the API wouldn't work properly in any case without createdb.js first). But the tests do make it a stronger argument.

File: API/models/model.go (hunk 1 above)
Comment (FrancoLiberali) [reply API/models/model.go:706]: Fixed in 6faa5fec

File: CLI/controllers/controller.go [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+var C = Controller{
+	API:     API,
+	Ogree3D: Ogree3D,
+}
+
+type Controller struct {
+	API     APIPort
+	Ogree3D Ogree3DPort
+}
```
Comment (GeremWD) [inline CLI/controllers/controller.go:11]: Im not a big fan of these extra imbricated global variables, it is hard to follow. I should have made the remark for the unity connection PR (because of the Ogree3DConnection one), but i did not pay much attention to it. 
We already have one big global variable for the state of the CLI. I dont see why the "isConnected" boolean of Ogree3DConnection would not be part of that state. Here APIPort does not have any member so I do not see the point either. We can make new packages if we want to the functions to be better organized.

File: CLI/astutil.go [hunk 11]
```diff
@@ -65,6 +97,20 @@ func nodeToVec(n node, size int, name string) ([]float64, error) {
...
+	}
+
+	color, ok := utils.ValToColor(colorInf)
+	if !ok {
+		return "", fmt.Errorf("Please provide a valid 6 digit Hex value for the color")
+	}
+
+	return color, nil
+}
+
```
Comment (GeremWD) [inline CLI/astutil.go:113]: I had implemented similar checks before but we decided with helder that it was better to remove them because it is supposed to be implemented on the API side already, so it would make it harder to maintain because you have to keep both sides in sync.

File: API/models/model.go (hunk 1 above)
Comment (helderbetiol) [reply API/models/model.go:706]: ok for the tag that does not exist, but it accepts "tags" : "a string that is not an array", it does not add it to the object but it does create the object and gives a 201 when an error of format validation was expected.

File: CLI/controllers/controller.go (hunk 10 above)
Comment (FrancoLiberali) [reply CLI/controllers/controller.go:11]: All these changes are to ensure the testability of the cli. To be able to make tests that do not depend on the api, mocks are necessary, to be able to simulate the response that the api would give for certain request. You can see examples of this in the tests I am adding in delete_test.go and update_test.go. In future pull requests I am adding many more tests. It is necessary to move from function schema to structures because a function call is impossible to mock. Instead, the functions are passed to structures with methods and an interface is created with all the methods of that structure, to be able to create the mocks (see interface APIPort and mocks/controllers/APIPort.go). The mocks are generated automatically with mockery.
Anyway, the global variables should not be necessary, since the objects could create their dependencies (and use dependency injection in case of complexity), but the problem is that to achieve this you have to pass all the functions to the new schema and it is a change that will have to be done progressively.
On the other hand, you could avoid global variables using a singleton pattern, but I think that the complexity of the application does not merit doing that work (there are only 3 global variables).

File: CLI/astutil.go (hunk 11 above)
Comment (FrancoLiberali) [reply CLI/astutil.go:113]: I don't quite understand this comment because it is in the line that says "return color, nil" and I suspect you were referring to another part of the code.

File: API/repository/base.go [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+	ctx, cancel := u.Connect()
+	defer cancel()
+
+	_, err := CreateObject(ctx, u.EntityToString(u.DOMAIN), map[string]any{
+		"id":       tenantName,
+		"name":     tenantName,
+		"category": "domain",
+		"attributes": map[string]any{
+			"color":       "ffffff",
+			"description": []any{},
```
[the remark this answers is not in the document]
Comment (FrancoLiberali) [reply API/repository/base.go]: fixed in f1a74306

File: API/models/model.go (hunk 1 above)
Comment (FrancoLiberali) [reply API/models/model.go:706]: fixed in 54592f6c

File: CLI/astutil.go (hunk 11 above)
Comment (GeremWD) [reply CLI/astutil.go:113]: Sorry for the misselection, I was talking about the regex check right above for the slug. I thought you also added one for the color, but apparently there was already a function checking that the color is valid, but I think it should be removed because it is also checked on the API side.

File: CLI/controllers/controller.go (hunk 10 above)
Comment (GeremWD) [reply CLI/controllers/controller.go:11]: Ok I understand better now. Maybe it will be worth spending time trying to uniformize the pattern across the whole project, or find another way to mock functions. But I agree that it can be done progressively.

**Review by GeremWD — APPROVED** [review]

File: API/models/model.go (hunk 1 above)
Comment (helderbetiol) [reply API/models/model.go:706]: Another observation: if we send a list of tags and only one of them is not valid, the response is just "tag not found" and we don't know which tag.

File: CLI/astutil.go (hunk 11 above)
Comment (FrancoLiberali) [reply CLI/astutil.go:82]: Ok now i see. Fixed in caf10df1. Only now instead of having a clear error the error is:
Error: [Response From API] JSON body doesn't validate with the expected JSON schema
    /slug does not match pattern '^[a-z0-9-_]+$'

But I think it is a more general problem than with slugs, I will create an issue to improve the error message in these cases.

File: API/models/model.go (hunk 1 above)
Comment (FrancoLiberali) [reply API/models/model.go:706]: fixed in 5fa22073

**Review by helderbetiol — APPROVED** [review]
